weave/trace_server/calls_query_builder.py

```diff
@@ -0,0 +1,726 @@
+        for condition in self.query_conditions:
+            if condition.is_heavy():
+                outer_query.query_conditions.append(condition)
+            else:
+                filter_query.query_conditions.append(condition)
+
+        # Hardcoded Filter - always light
+        filter_query.hardcoded_filter = self.hardcoded_filter
+
+        # Order Fields:
+        if has_light_order_filter:
+            filter_query.order_fields = self.order_fields
+            filter_query.limit = self.limit
+            filter_query.offset = self.offset
+            # SUPER IMPORTANT: still need to re-sort the final query
+            outer_query.order_fields = self.order_fields
+        else:
+            outer_query.order_fields = self.order_fields
+            outer_query.limit = self.limit
+            outer_query.offset = self.offset
+
+        return f"""
+        WITH filtered_calls AS ({filter_query._as_sql_base_format(pb, table_alias)})
+        {outer_query._as_sql_base_format(pb, table_alias, id_subquery_name="filtered_calls")}
+        """
+
+    def _as_sql_base_format(
+        self,
+        pb: ParamBuilder,
+        table_alias: str,
+        id_subquery_name: typing.Optional[str] = None,
+    ) -> str:
+        select_fields_sql = ", ".join(
+            [field.as_select_sql(pb, table_alias) for field in self.select_fields]
+        )
+
+        having_filter_sql = ""
+        having_conditions_sql = []
+        if len(self.query_conditions) > 0:
+            having_conditions_sql.extend(
+                [c.as_sql(pb, table_alias) for c in self.query_conditions]
+            )
+        if self.hardcoded_filter is not None:
+            having_conditions_sql.append(self.hardcoded_filter.as_sql(pb, table_alias))
+
+        if len(having_conditions_sql) > 0:
+            having_filter_sql = "HAVING " + combine_conditions(
+                having_conditions_sql, "AND"
+            )
+
+        order_by_sql = ""
+        if len(self.order_fields) > 0:
+            order_by_sql = "ORDER BY " + ", ".join(
+                [
+                    order_field.as_sql(pb, table_alias)
+                    for order_field in self.order_fields
+                ]
+            )
+
+        limit_sql = ""
+        if self.limit is not None:
+            limit_sql = f"LIMIT {self.limit}"
+
+        offset_sql = ""
+        if self.offset is not None:
+            offset_sql = f"OFFSET {self.offset}"
+
+        id_subquery_sql = ""
+        if id_subquery_name is not None:
+            id_subquery_sql = f"AND (id IN {id_subquery_name})"
+
+        project_param = pb.add_param(self.project_id)
+
+        # Special Optimization
+        id_mask_sql = ""
+        if self.hardcoded_filter and self.hardcoded_filter.filter.call_ids:
+            id_mask_sql = f"AND (id IN {_param_slot(pb.add_param(self.hardcoded_filter.filter.call_ids), 'Array(String)')})"
+        # TODO: We should also pull out id-masks from the dynamic query
+
+        return f"""
```

Comment:
This string is going to have a bunch of extra whitespace in it - a possible future enhancement would be to tidy it up in case we print or log it somewhere.

Reply:
agreed - in the test suite i use a formatter for consistency but didn't want to put that in the hot path.

Reply:
added a safe formatter